src/apps/containers.clj

```diff
@@ -585,34 +599,210 @@
     (if-not (nil? container-info)
       {:container_volumes_from (:container_volumes_from container-info)})))
 
+(defn ports
+  "Returns the port information for the given container_settings uuid."
+  [settings-uuid]
+  (select container-ports (where {:container_settings_id (uuidify settings-uuid)})))
+
+(defn port
+  "Returns a specific port indicated by the port UUID."
+  [port-uuid]
+  (first (select container-ports (where {:id (uuidify port-uuid)}))))
+
+(defn port?
+  "Returns true if the port exists."
+  [port-uuid]
+  (pos? (count (select container-ports (where {:id (uuidify port-uuid)})))))
+
+(defn port-mapping?
+  "Returns true if the the specific combination of fields exists in the
+   database."
+  [settings-uuid host-port container-port bind-to-host]
+  (pos? (count (select container-ports
+                      (where (and (= :container_settings_id (uuidify settings-uuid))
+                                  (= :host_port host-port)
+                                  (= :container_port container-port)
+                                  (= :bind_to_host bind-to-host)))))))
+
+(defn port-mapping
+  [settings-uuid host-port container-port bind-to-host]
+  (first (select container-ports
+                 (where (and (= :container_settings_id (uuidify settings-uuid))
+                             (= :host_port host-port)
+                             (= :container_port container-port)
+                             (= :bind_to_host bind-to-host))))))
+
+(defn settings-has-port?
+  [settings-uuid port-uuid]
+  (pos? (count (select container-ports
+                       (where {:container_settings_id (uuidify settings-uuid)
+                               :id                    (uuidify port-uuid)})))))
+
+(defn add-port
+  [settings-uuid port-map]
+  (let [host-port      (:host_port port-map)
+        container-port (:container_port port-map)
+        bind-to-host   (:bind_to_host port-map)]
+    (if (port-mapping? settings-uuid host-port container-port bind-to-host)
+      (throw (Exception. (str "port mapping already exists: " settings-uuid " " host-port " " container-port " " bind-to-host))))
```

Comment:
I think this will cause apps to return a 500 status, but a 400 is probably what you want. There's a set of helper functions in `clojure-commons.exception-util` that may help. (I normally use `cxu` as tis namespace alias.) In this case, you can call `cxu/exists` to have the service return a 400 status code along with a response body containing an error code of `ERR_EXISTS`:

```clojure
    (if (port-mapping? settings-uuid host-port container-port bind-to-host)
      (cxu/exists (str "port mapping already exists: " settings-uuid " " host-port " " container-port " " bind-to host)))
```